byzcoin/service.go

```diff
@@ -1340,7 +1355,7 @@ func (s *Service) updateTrieCallback(sbID skipchain.SkipBlockID) error {
 	// Check if the polling needs to be updated.
 	s.pollChanMut.Lock()
 	scIDstr := string(sb.SkipChainID())
-	if nodeIsLeader {
+	if nodeIsLeader && !s.catchingUp {
```

Comment:
if not mistaken, the `s.catchingUp` flag is locked/unlocked by `s.updateTrieLock` instead of `s.catchingLock`, which seems a bit confusing. i'm wondering whether we can refactor the catchup logic into its own go-routine and the byzcoin service will interact with it only using channels. i'm not saying we should change anything now, just throwing the idea out there.

Reply:
The problem here is that the callback after a block is stored is called during a catch up (it has to) but some actions like starting the heartbeat/monitor should not be trigger until the end of the catch up. So you need a way to tell if the callback is called in a catch up context, and I'm not sure that channels could solve this.